### PR TITLE
publikator: Repo search in cache

### PR DESCRIPTION
**Pull request by patrickvenetz**
This Pull Requests enables us to query available repositories without hitting GitHub API, the main objective.

It assumes, repositories are only altered via Publikator API. (However, if data is altered on GitHub itself, repo index can be rebuild.)

Repository data provided in a [new ElasticSearch index, named `repo`](https://github.com/orbiting/backends/pull/98/files#diff-24a173a602d6889eabfdeeac7c59a084R5).

Code is split into to parts:
a) Searching repository cache data, see [servers/publikator/lib/cache/search.js](https://github.com/orbiting/backends/pull/98/files#diff-58c3671607d5e7355c84fee5b4c7198d)
b) Indexing (and upserting) repository cache data, see [servers/publikator/lib/cache/upsert.js](https://github.com/orbiting/backends/pull/98/files#diff-fd49690c6dc88b75354b8005c5e3472d)

A new query named `search` exposes this cached repository data. It imitates already existing `repos` query which makes it an excellent drop-in replacement:

`repos`, hitting GitHub API hard:

```gql
{ repos(first: 10, search: "bananas") {
    totalCount
} }
```

`search` as named query, using ElasticSearch cache:

```gql
{ repos: search(first: 10, search: "bananas") {
    totalCount
} }
```

Caveats: It adds [an additional `template` argument](https://github.com/orbiting/backends/pull/98/files#diff-51f515cd2bdcb767e2e53edff74b05bbR32) (to narrow returned repos to e.g. "format") and did not implemented [`last`](https://github.com/orbiting/backends/pull/98/files#diff-51f515cd2bdcb767e2e53edff74b05bbR27) argument (impractical and so far of little use).

Multiple Publikator mutations will upsert repository data:
- [`commit`](https://github.com/orbiting/backends/pull/98/files#diff-59f8b60e66d68d95c1703086f35d0fc5)
- [`editRepoMeta`](https://github.com/orbiting/backends/pull/98/files#diff-8b653f23066c93cf9d890d4c22f6c558)
- [`placeMilestone`](https://github.com/orbiting/backends/pull/98/files#diff-1ee0b62c42f58fc32d5455d30f043fac)
- [`publish`](https://github.com/orbiting/backends/pull/98/files#diff-6e358742c5bc46972ccb65ea7e02ad7f)
- [`removeMilestone`](https://github.com/orbiting/backends/pull/98/files#diff-85664a2c894d52f5e1599da17c1ad3c4)

Initial indexing, and later rebuilding, can be done via usual pull script:
```sh
yarn pull:elasticsearch --flush --index repo
```

Other implementation notes:
- [Mapping is build dynamically](https://github.com/orbiting/backends/pull/98/files#diff-3ff0d59f44d91f6135d2c907c8a2b644R9). There is currently no need to "nail" mapping.
- Introduced [`searchable` flag](https://github.com/orbiting/backends/pull/98/files#diff-02178489914e37b5110bde1402264cd9R327) on indices. If `true`, and index is searchable via Republik `search` query.
- Fix to allow users with editor role [to retrieve documents via document IDs](https://github.com/orbiting/backends/pull/98/files#diff-53007a5a8047cfbb697e25b099cfed17R100). Before only a single ID worked.
- Removed [useless schema setting in `packages/search/lib/Documents.js`](https://github.com/orbiting/backends/pull/98/files#diff-f27006a98b088509a7149a3987e29453L66) as "__state" is not a term nor can it be aggregated that way.

Here [is a series of manual tests](https://docs.google.com/spreadsheets/d/1js_y3Ocy-ck33I9pP16hID6jB5W1a727M94YCMES4qs/edit#gid=0) to ensure `search` query (new) responds with same or close to similar results than `repos` did.

### CODE REVIEW

File: packages/search/lib/indices/repos.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const type = 'Repo'
+
+module.exports = {
+  type,
+  name: type.toLowerCase(),
+  searchable: false,
```
Comment (tpreusse) [inline packages/search/lib/indices/repos.js]: Odd naming. Maybe just explicitly define which index you want to search in `search/graphql/resolvers/_queries/search.js` or expose a default list in `search/lib/indices/index.js` without repos?

File: servers/publikator/lib/cache/search.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    'contentMeta.facebookTitle',
+    'contentMeta.format',
+    'contentMeta.seriesMaster.episodes.document',
+    'contentMeta.seriesMaster.episodes.label',
+    'contentMeta.seriesMaster.episodes.title',
+    'contentMeta.seriesMaster.title',
+    'contentMeta.slug',
+    'contentMeta.subject',
+    'contentMeta.template',
+    'contentMeta.title',
```
[the remark this answers is not in the document]
Comment (patrickvenetz) [reply servers/publikator/lib/cache/search.js]: > (are those linked episodes document contents?)

No.

File: packages/search/lib/indices/repos.js (hunk 1 above)
Comment (patrickvenetz) [reply packages/search/lib/indices/repos.js]: Understand, yet sticking to it.

File: servers/publikator/graphql/schema.js [hunk 3]
```diff
@@ -17,6 +17,21 @@ type queries {
   ): RepoConnection!
   repo(id: ID!): Repo!
   embed(id: ID!, embedType: EmbedType!): Embed!
+
+  """
+  This query is a cached version of repos query. It uses cached information
+  about repositories.
+  """
+  search(
```
[the remark this answers is not in the document]
Comment (patrickvenetz) [reply servers/publikator/graphql/schema.js]: Renamed to `reposSearch`.

**Comment by patrickvenetz** [comment]
@tpreusse Be so kind an revisit it. Adopted most of your good review comments.

File: servers/publikator/lib/cache/search.js [hunk 4]
```diff
@@ -13,7 +13,7 @@ const client = elasticsearch.client()
 const getSort = (args) => {
   // Default sorting
   if (!args.orderBy) {
-    return { sort: { 'latestCommit.date': 'desc' } }
+    return
```
Comment (tpreusse) [inline servers/publikator/lib/cache/search.js]: `args.search ? undefined : { sort: { 'latestCommit.date': 'desc' } }`, IMO you still want `'latestCommit.date': 'desc'` when using it without search (list on publikator index page)

**Pull request merged**